File: franka_simulink_library/blocks/franka_torque_control.m
function franka_torque_control(block)
%FRANKA_TORQUE_CONTROL Level-2 MATLAB S-Function for Franka robot torque control
%  
%  This block wraps libfranka's robot.control() with torque callback.
%  It outputs a function-call signal to trigger an external controller subsystem.
%
%  Inputs:
%    1. Enable    (1x1) - Rising edge starts control, falling edge stops
%    2. tau_J_d   (7x1) - Commanded joint torques (from controller subsystem)
%
%  Outputs:
%    1. fcall     (function-call) - Triggers controller subsystem at 1kHz
%    2. q         (7x1) - Measured joint positions
%    3. dq        (7x1) - Measured joint velocities
%    4. status    (1x1) - Control status: 0=Idle, 1=Running, 2=Error
%
%  Parameters:
%    1. robot_ip  (string) - Robot IP address
%
%  Copyright (c) 2025 Franka Robotics GmbH

setup(block);

%% ========================================================================
function setup(block)

    % Register number of dialog parameters
    block.NumDialogPrms = 1;  % robot_ip
    block.DialogPrmsTunable = {'Nontunable'};
    
    % Register number of input ports
    block.NumInputPorts = 2;
    
    % Register number of output ports  
    % Output 1 MUST be the function-call (Simulink requirement)
    block.NumOutputPorts = 4;
    
    % Setup input port properties
    % Port 1: Enable
    block.InputPort(1).Dimensions = 1;
    block.InputPort(1).DatatypeID = 0;  % double
    block.InputPort(1).Complexity = 'Real';
    block.InputPort(1).DirectFeedthrough = true;
    block.InputPort(1).SamplingMode = 'Sample';
    
    % Port 2: tau_J_d (joint torques)
    block.InputPort(2).Dimensions = 7;
    block.InputPort(2).DatatypeID = 0;  % double
    block.InputPort(2).Complexity = 'Real';
    block.InputPort(2).DirectFeedthrough = true;
    block.InputPort(2).SamplingMode = 'Sample';
    
    % Setup output port properties
    % Port 1: Function-call output (MUST be first!)
    block.OutputPort(1).Dimensions = 1;
    block.OutputPort(1).DatatypeID = 0;  % Will be overridden to fcn_call in TLC
    block.OutputPort(1).Complexity = 'Real';
    block.OutputPort(1).SamplingMode = 'Sample';
    
    % Port 2: q (joint positions)
    block.OutputPort(2).Dimensions = 7;
    block.OutputPort(2).DatatypeID = 0;  % double
    block.OutputPort(2).Complexity = 'Real';
    block.OutputPort(2).SamplingMode = 'Sample';
    
    % Port 3: dq (joint velocities)
    block.OutputPort(3).Dimensions = 7;
    block.OutputPort(3).DatatypeID = 0;  % double
    block.OutputPort(3).Complexity = 'Real';
    block.OutputPort(3).SamplingMode = 'Sample';
    
    % Port 4: status
    block.OutputPort(4).Dimensions = 1;
    block.OutputPort(4).DatatypeID = 0;  % double
    block.OutputPort(4).Complexity = 'Real';
    block.OutputPort(4).SamplingMode = 'Sample';
    
    % Register sample times
    % [0.001 0] = 1kHz discrete sample time
    block.SampleTimes = [0.001 0];
    
    % Specify block simStateCompliance
    block.SimStateCompliance = 'DefaultSimState';
    
    % Register methods
    block.RegBlockMethod('CheckParameters', @CheckPrms);
    block.RegBlockMethod('PostPropagationSetup', @DoPostPropSetup);
    block.RegBlockMethod('Start', @Start);
    block.RegBlockMethod('Outputs', @Outputs);
    block.RegBlockMethod('Terminate', @Terminate);
    block.RegBlockMethod('WriteRTW', @WriteRTW);

%% ========================================================================
function CheckPrms(block)
    % Validate robot_ip parameter
    robot_ip = block.DialogPrm(1).Data;
    if ~ischar(robot_ip) && ~isstring(robot_ip)
        error('Robot IP must be a string');
    end

%% ========================================================================
function DoPostPropSetup(block)
    % Setup DWork vectors for state management
    block.NumDworks = 2;
    
    % DWork 1: Previous enable value (for edge detection)
    block.Dwork(1).Name = 'PrevEnable';
    block.Dwork(1).Dimensions = 1;
    block.Dwork(1).DatatypeID = 0;  % double
    block.Dwork(1).Complexity = 'Real';
    block.Dwork(1).UsedAsDiscState = true;
    
    % DWork 2: Control state (0=idle, 1=running, 2=error)
    block.Dwork(2).Name = 'ControlState';
    block.Dwork(2).Dimensions = 1;
    block.Dwork(2).DatatypeID = 0;  % double
    block.Dwork(2).Complexity = 'Real';
    block.Dwork(2).UsedAsDiscState = true;

%% ========================================================================
function Start(block)
    % Initialize DWork
    block.Dwork(1).Data = 0;  % PrevEnable = 0
    block.Dwork(2).Data = 0;  % ControlState = Idle

%% ========================================================================
function Outputs(block)
    % This is a placeholder for simulation
    % Actual implementation is in TLC for code generation
    
    % For simulation, just pass through zeros
    block.OutputPort(2).Data = zeros(7, 1);  % q
    block.OutputPort(3).Data = zeros(7, 1);  % dq
    block.OutputPort(4).Data = 0;            % status

%% ========================================================================
function Terminate(block)
    % Cleanup (placeholder for simulation)

%% ========================================================================
function WriteRTW(block)
    % Write parameters to RTW file for code generation
    
    robot_ip = block.DialogPrm(1).Data;
    
    % Create robot_id from IP (remove dots)
    robot_id = strrep(robot_ip, '.', '');
    
    % Quote strings for TLC
    robot_ip_quoted = sprintf('''%s''', robot_ip);
    robot_id_quoted = sprintf('''%s''', robot_id);
    
    % Write parameters
    block.WriteRTWParam('string', 'robot_ip', robot_ip_quoted);
    block.WriteRTWParam('string', 'robot_id', robot_id_quoted);

File: franka_simulink_library/blocks/franka_torque_control.cpp
/*
 * franka_torque_control.cpp - C++ Level-2 S-Function for Franka robot torque control
 *
 * This block wraps libfranka's robot.control() with torque callback.
 * It outputs a function-call signal to trigger an external controller subsystem.
 *
 * Inputs:
 *   0. Enable    (1x1)   - Rising edge starts control, falling edge stops
 *   1. robot_ip  (16x1)  - Robot IP as ASCII chars (use String Constant + String to ASCII)
 *   2. tau_J_d   (7x1)   - Commanded joint torques (read from controller subsystem)
 *
 * Outputs:
 *   0. fcall     (function-call) - Triggers controller subsystem at 1kHz [MUST BE FIRST]
 *   1. q         (7x1) - Measured joint positions
 *   2. dq        (7x1) - Measured joint velocities
 *
 * Copyright (c) 2025 Franka Robotics GmbH
 */

#define S_FUNCTION_NAME  franka_torque_control
#define S_FUNCTION_LEVEL 2

#include "simstruc.h"

#include <string>
#include <cstring>

/* No parameters required - all inputs come from ports */
#define NUM_PARAMS     0

/* Input port indices */
#define IN_ENABLE     0
#define IN_ROBOT_IP   1
#define IN_TAU_J_D    2
#define NUM_INPUTS    3

/* Output port indices */
#define OUT_FCALL     0   /* Function-call output - MUST BE FIRST */
#define OUT_Q         1
#define OUT_DQ        2
#define NUM_OUTPUTS   3

/* DWork indices */
#define DWORK_PREV_ENABLE 0
#define NUM_DWORK         1

/* No parameters to check - all inputs come from ports */

/* ========================================================================
 * mdlInitializeSizes - Initialize block sizes
 * ======================================================================== */
static void mdlInitializeSizes(SimStruct *S)
{
    /* No parameters - all inputs come from ports */
    ssSetNumSFcnParams(S, NUM_PARAMS);
    
#if defined(MATLAB_MEX_FILE)
    if (ssGetNumSFcnParams(S) != ssGetSFcnParamsCount(S)) {
        return; /* Parameter mismatch reported by Simulink */
    }
#endif
    
    /* States */
    ssSetNumContStates(S, 0);
    ssSetNumDiscStates(S, 0);
    
    /* ====================================================================
     * INPUT PORTS
     * ==================================================================== */
    if (!ssSetNumInputPorts(S, NUM_INPUTS)) return;
    
    /* Port 0: Enable (1x1) */
    ssSetInputPortWidth(S, IN_ENABLE, 1);
    ssSetInputPortDataType(S, IN_ENABLE, SS_DOUBLE);
    ssSetInputPortDirectFeedThrough(S, IN_ENABLE, 1);
    ssSetInputPortRequiredContiguous(S, IN_ENABLE, 1);
    
    /* Port 1: robot_ip (16 ASCII chars, null-terminated) */
    ssSetInputPortWidth(S, IN_ROBOT_IP, 16);
    ssSetInputPortDataType(S, IN_ROBOT_IP, SS_UINT8);
    ssSetInputPortDirectFeedThrough(S, IN_ROBOT_IP, 1);
    ssSetInputPortRequiredContiguous(S, IN_ROBOT_IP, 1);
    
    /* Port 2: tau_J_d (7x1) */
    ssSetInputPortWidth(S, IN_TAU_J_D, 7);
    ssSetInputPortDataType(S, IN_TAU_J_D, SS_DOUBLE);
    ssSetInputPortDirectFeedThrough(S, IN_TAU_J_D, 1);
    ssSetInputPortRequiredContiguous(S, IN_TAU_J_D, 1);
    
    /* ====================================================================
     * OUTPUT PORTS
     * ==================================================================== */
    if (!ssSetNumOutputPorts(S, NUM_OUTPUTS)) return;
    
    /* Port 0: Function-call output (MUST BE FIRST!) */
    ssSetOutputPortWidth(S, OUT_FCALL, 1);
    ssSetOutputPortDataType(S, OUT_FCALL, SS_FCN_CALL);
    
    /* Port 1: q (7x1) */
    ssSetOutputPortWidth(S, OUT_Q, 7);
    ssSetOutputPortDataType(S, OUT_Q, SS_DOUBLE);
    
    /* Port 2: dq (7x1) */
    ssSetOutputPortWidth(S, OUT_DQ, 7);
    ssSetOutputPortDataType(S, OUT_DQ, SS_DOUBLE);
    
    /* ====================================================================
     * SAMPLE TIME
     * ==================================================================== */
    ssSetNumSampleTimes(S, 1);
    
    /* ====================================================================
     * WORK VECTORS
     * ==================================================================== */
    ssSetNumRWork(S, 0);
    ssSetNumIWork(S, 0);
    ssSetNumPWork(S, 0);
    ssSetNumModes(S, 0);
    ssSetNumNonsampledZCs(S, 0);
    
    /* DWork for edge detection */
    ssSetNumDWork(S, NUM_DWORK);
    ssSetDWorkWidth(S, DWORK_PREV_ENABLE, 1);
    ssSetDWorkDataType(S, DWORK_PREV_ENABLE, SS_DOUBLE);
    ssSetDWorkName(S, DWORK_PREV_ENABLE, "PrevEnable");
    
    /* Options */
    ssSetSimStateCompliance(S, USE_DEFAULT_SIM_STATE);
    ssSetOptions(S, SS_OPTION_CALL_TERMINATE_ON_EXIT);
}

/* ========================================================================
 * mdlInitializeSampleTimes - Initialize sample times
 * ======================================================================== */
static void mdlInitializeSampleTimes(SimStruct *S)
{
    /* 1kHz discrete sample time */
    ssSetSampleTime(S, 0, 0.001);
    ssSetOffsetTime(S, 0, 0.0);
    ssSetModelReferenceSampleTimeDefaultInheritance(S);
    
    /* Register function-call output port */
    ssSetCallSystemOutput(S, OUT_FCALL);
}

/* ========================================================================
 * mdlStart - Initialize
 * ======================================================================== */
#define MDL_START
#if defined(MDL_START)
static void mdlStart(SimStruct *S)
{
    /* Initialize DWork */
    real_T *prevEnable = (real_T*)ssGetDWork(S, DWORK_PREV_ENABLE);
    *prevEnable = 0.0;
}
#endif

/* ========================================================================
 * mdlOutputs - Compute outputs
 * ======================================================================== */
static void mdlOutputs(SimStruct *S, int_T tid)
{
    /* For simulation only - actual implementation is in TLC */
    real_T *q  = ssGetOutputPortRealSignal(S, OUT_Q);
    real_T *dq = ssGetOutputPortRealSignal(S, OUT_DQ);
    
    /* Output zeros in simulation */
    for (int i = 0; i < 7; i++) {
        q[i]  = 0.0;
        dq[i] = 0.0;
    }
    
    UNUSED_ARG(tid);
}

/* ========================================================================
 * mdlTerminate - Cleanup
 * ======================================================================== */
static void mdlTerminate(SimStruct *S)
{
    UNUSED_ARG(S);
}

/* ========================================================================
 * mdlRTW - Write parameters to RTW file for code generation
 * ======================================================================== */
#if defined(MATLAB_MEX_FILE)
#define MDL_RTW
static void mdlRTW(SimStruct *S)
{
    /* robot_ip comes from input port - no parameters to write */
    UNUSED_ARG(S);
}
#endif

/* ========================================================================
 * Required S-function trailer
 * ======================================================================== */
#ifdef MATLAB_MEX_FILE
#include "simulink.c"
#else
#include "cg_sfun.h"
#endif

